File: examples/school/ts/src/handlers.ts
import {
  BusinessError, ErrBodyIsBroken, ErrPageIsInvalid, ErrPathIsBroken,
  checkStudentRecord, type Student, type CourseID, type StudentID,
} from "./domain.js";
import {
  buildCourseRecord, buildStudentRecord, renderCourseView, renderStudentView,
} from "./dto.js";
import type { CourseStore, Page, StudentStore, TokenIssuer } from "./providers.js";
import type { Request, Response, Route } from "./transport.js";

// SchoolAPI Holds the Cast every Story Needs.
// Three Providers, no Libraries: a Test can Hand it three Fakes.
export class SchoolAPI {
  constructor(
    private readonly students: StudentStore,
    private readonly courses: CourseStore,
    private readonly tokens: TokenIssuer,
  ) {}

  // declareSchoolRoutes is the Libretto.
  // Read it once and you Know the whole Service.
  declareSchoolRoutes(): Route[] {
    return [
      { method: "POST", pattern: "/token", handle: (r) => this.mintAccessToken(r) },

      { method: "GET", pattern: "/students", handle: (r) => this.listStudentRecords(r) },
      { method: "POST", pattern: "/students", handle: (r) => this.addStudentRecord(r) },
      { method: "GET", pattern: "/students/{id}", handle: (r) => this.showStudentRecord(r) },
      { method: "PUT", pattern: "/students/{id}", handle: (r) => this.saveStudentRecord(r) },
      { method: "DELETE", pattern: "/students/{id}", handle: (r) => this.dropStudentRecord(r) },

      { method: "GET", pattern: "/courses", handle: (r) => this.listCourseRecords(r) },
      { method: "POST", pattern: "/courses", handle: (r) => this.addCourseRecord(r) },
      { method: "GET", pattern: "/courses/{id}", handle: (r) => this.showCourseRecord(r) },
    ];
  }

  // mintAccessToken Hands out a Token, and Guards nothing else.
  mintAccessToken(_: Request): Response {
    return attempt(() => ({
      status: 201,
      body: { token: this.tokens.issueAccessToken("student-registry") },
    }));
  }

  // listStudentRecords Tells one Page of the Roll.
  listStudentRecords(request: Request): Response {
    return attempt(() => ({
      status: 200,
      body: this.students.selectStudentPage(readPageRequest(request)).map(renderStudentView),
    }));
  }

  // showStudentRecord Tells the Story of one Student.
  showStudentRecord(request: Request): Response {
    return attempt(() => ({
      status: 200,
      body: renderStudentView(this.students.selectStudentRow(readPathNumber(request) as StudentID)),
    }));
  }

  // addStudentRecord Enrols someone new.
  addStudentRecord(request: Request): Response {
    return attempt(() => ({
      status: 201,
      body: renderStudentView(this.students.insertStudentRow(this.readStudentBody(request, 0))),
    }));
  }

  // saveStudentRecord Rewrites an Enrolment that already Exists.
  saveStudentRecord(request: Request): Response {
    return attempt(() => {
      const student = this.readStudentBody(request, readPathNumber(request));

      return { status: 200, body: renderStudentView(this.students.updateStudentRow(student)) };
    });
  }

  // dropStudentRecord Ends an Enrolment.
  dropStudentRecord(request: Request): Response {
    return attempt(() => {
      this.students.deleteStudentRow(readPathNumber(request) as StudentID);

      return { status: 204, body: null };
    });
  }

  // listCourseRecords Tells one Page of the Catalogue.
  listCourseRecords(request: Request): Response {
    return attempt(() => ({
      status: 200,
      body: this.courses.selectCoursePage(readPageRequest(request)).map(renderCourseView),
    }));
  }

  // showCourseRecord Tells the Story of one Course.
  showCourseRecord(request: Request): Response {
    return attempt(() => ({
      status: 200,
      body: renderCourseView(this.courses.selectCourseRow(readPathNumber(request) as CourseID)),
    }));
  }

  // addCourseRecord Opens a new Course.
  addCourseRecord(request: Request): Response {
    return attempt(() => ({
      status: 201,
      body: renderCourseView(this.courses.insertCourseRow(buildCourseRecord(readJsonBody(request), 0))),
    }));
  }

  // readStudentBody Decodes the Wire, Validates it, then Checks the Course.
  private readStudentBody(request: Request, id: number): Student {
    const student = buildStudentRecord(readJsonBody(request), id);

    const broken = checkStudentRecord(student);
    if (broken) throw broken;

    this.courses.selectCourseRow(student.course);

    return student;
  }
}

// attempt is the one Place that Turns a thrown Failure into a Reply.
// The Mapping Lives here once, never inside a Handler.
function attempt(story: () => Response): Response {
  try {
    return story();
  } catch (failure) {
    return buildFailureReply(failure);
  }
}

// buildFailureReply Maps a Business Failure onto an HTTP Code.
export function buildFailureReply(failure: unknown): Response {
  if (failure instanceof BusinessError) {
    const status = { invalid: 400, absent: 404, taken: 409 }[failure.kind];

    return { status, body: { error: failure.message } };
  }

  return { status: 500, body: { error: "unexpected Failure" } };
}

// readJsonBody Refuses anything the Wire Cannot Parse.
function readJsonBody<T>(request: Request): T {
  try {
    return JSON.parse(request.body || "{}") as T;
  } catch {
    throw ErrBodyIsBroken;
  }
}

// readPathNumber Pulls an Identity out of the Path.
function readPathNumber(request: Request): number {
  const raw = Number(request.path["id"]);

  if (!Number.isInteger(raw) || raw < 0) throw ErrPathIsBroken;

  return raw;
}

// readPageRequest Reads Pagination, Defaulting to the whole Set.
function readPageRequest(request: Request): Page {
  const page = { number: Number(request.query["page"] ?? 0), size: Number(request.query["size"] ?? 0) };

  if (page.number < 0 || page.size < 0) throw ErrPageIsInvalid;

  return page;
}

File: examples/school/go/handlers.go
package main

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
)

// SchoolAPI Holds the Cast every Story Needs.
// Three Providers, no Libraries: a Test can Hand it three Fakes.
type SchoolAPI struct {
	Students StudentStore
	Courses  CourseStore
	Tokens   TokenIssuer
}

// DeclareSchoolRoutes is the Libretto.
// Read it once and you Know the whole Service.
func (a SchoolAPI) DeclareSchoolRoutes() []Route {
	return []Route{
		{"POST", "/token", a.MintAccessToken},

		{"GET", "/students", a.ListStudentRecords},
		{"POST", "/students", a.AddStudentRecord},
		{"GET", "/students/{id}", a.ShowStudentRecord},
		{"PUT", "/students/{id}", a.SaveStudentRecord},
		{"DELETE", "/students/{id}", a.DropStudentRecord},

		{"GET", "/courses", a.ListCourseRecords},
		{"POST", "/courses", a.AddCourseRecord},
		{"GET", "/courses/{id}", a.ShowCourseRecord},
	}
}

// MintAccessToken Hands out a Token, and Guards nothing else.
func (a SchoolAPI) MintAccessToken(req Request) Response {
	token, err := a.Tokens.IssueAccessToken("student-registry")
	if err != nil {
		return BuildFailureReply(err)
	}

	return Response{http.StatusCreated, map[string]string{"token": token}}
}

// ListStudentRecords Tells one Page of the Roll.
func (a SchoolAPI) ListStudentRecords(req Request) Response {
	page, err := ReadPageRequest(req)
	if err != nil {
		return BuildFailureReply(err)
	}

	students, err := a.Students.SelectStudentPage(page)
	if err != nil {
		return BuildFailureReply(err)
	}

	return Response{http.StatusOK, RenderStudentViews(students)}
}

// ShowStudentRecord Tells the Story of one Student.
func (a SchoolAPI) ShowStudentRecord(req Request) Response {
	id, err := ReadPathNumber(req)
	if err != nil {
		return BuildFailureReply(err)
	}

	student, err := a.Students.SelectStudentRow(StudentID(id))

	return BuildStudentReply(http.StatusOK, student, err)
}

// AddStudentRecord Enrols someone new.
func (a SchoolAPI) AddStudentRecord(req Request) Response {
	student, err := a.ReadStudentBody(req, 0)
	if err != nil {
		return BuildFailureReply(err)
	}

	stored, err := a.Students.InsertStudentRow(student)

	return BuildStudentReply(http.StatusCreated, stored, err)
}

// SaveStudentRecord Rewrites an Enrolment that already Exists.
func (a SchoolAPI) SaveStudentRecord(req Request) Response {
	id, err := ReadPathNumber(req)
	if err != nil {
		return BuildFailureReply(err)
	}

	student, err := a.ReadStudentBody(req, StudentID(id))
	if err != nil {
		return BuildFailureReply(err)
	}

	stored, err := a.Students.UpdateStudentRow(student)

	return BuildStudentReply(http.StatusOK, stored, err)
}

// DropStudentRecord Ends an Enrolment.
func (a SchoolAPI) DropStudentRecord(req Request) Response {
	id, err := ReadPathNumber(req)
	if err != nil {
		return BuildFailureReply(err)
	}

	if err := a.Students.DeleteStudentRow(StudentID(id)); err != nil {
		return BuildFailureReply(err)
	}

	return Response{http.StatusNoContent, nil}
}

// ListCourseRecords Tells one Page of the Catalogue.
func (a SchoolAPI) ListCourseRecords(req Request) Response {
	page, err := ReadPageRequest(req)
	if err != nil {
		return BuildFailureReply(err)
	}

	courses, err := a.Courses.SelectCoursePage(page)
	if err != nil {
		return BuildFailureReply(err)
	}

	return Response{http.StatusOK, RenderCourseViews(courses)}
}

// ShowCourseRecord Tells the Story of one Course.
func (a SchoolAPI) ShowCourseRecord(req Request) Response {
	id, err := ReadPathNumber(req)
	if err != nil {
		return BuildFailureReply(err)
	}

	course, err := a.Courses.SelectCourseRow(CourseID(id))
	if err != nil {
		return BuildFailureReply(err)
	}

	return Response{http.StatusOK, RenderCourseView(course)}
}

// AddCourseRecord Opens a new Course.
func (a SchoolAPI) AddCourseRecord(req Request) Response {
	var body CourseBody
	if err := json.Unmarshal(req.Body, &body); err != nil {
		return BuildFailureReply(ErrBodyIsBroken)
	}

	stored, err := a.Courses.InsertCourseRow(body.BuildCourseRecord(0))
	if err != nil {
		return BuildFailureReply(err)
	}

	return Response{http.StatusCreated, RenderCourseView(stored)}
}

// ReadStudentBody Decodes the Wire, Validates it, then Checks the Course.
func (a SchoolAPI) ReadStudentBody(req Request, id StudentID) (Student, error) {
	var body StudentBody
	if err := json.Unmarshal(req.Body, &body); err != nil {
		return Student{}, ErrBodyIsBroken
	}

	student := body.BuildStudentRecord(id)
	if err := student.CheckStudentRecord(); err != nil {
		return Student{}, err
	}

	_, err := a.Courses.SelectCourseRow(student.Course)

	return student, err
}

// ReadPathNumber Pulls an Identity out of the Path.
func ReadPathNumber(req Request) (uint64, error) {
	raw, err := strconv.ParseUint(req.Path["id"], 10, 64)
	if err != nil {
		return 0, ErrPathIsBroken
	}

	return raw, nil
}

// ReadPageRequest Reads Pagination, Defaulting to the whole Set.
func ReadPageRequest(req Request) (Page, error) {
	number, _ := strconv.Atoi(req.Query["page"])
	size, _ := strconv.Atoi(req.Query["size"])

	page := Page{Number: number, Size: size}

	return page, page.CheckPageBounds()
}

// BuildStudentReply Answers with a Student, or with why there is none.
func BuildStudentReply(code int, s Student, err error) Response {
	if err != nil {
		return BuildFailureReply(err)
	}

	return Response{code, RenderStudentView(s)}
}

// BuildFailureReply Maps a Business Failure onto an HTTP Code.
// The Mapping Lives here once, never inside a Handler.
func BuildFailureReply(err error) Response {
	body := map[string]string{"error": err.Error()}

	if errors.Is(err, ErrStudentUnknown) || errors.Is(err, ErrCourseUnknown) {
		return Response{http.StatusNotFound, body}
	}

	if errors.Is(err, ErrRutTaken) {
		return Response{http.StatusConflict, body}
	}

	return Response{http.StatusBadRequest, body}
}
